README.md
# QSAR-Report
QMRF and QPRF reporting extension to OpenTox ruby modules and lazar.
## About
Class for QMRF and QPRF reporting.  
* QMRF:
  Provides a ruby OpenTox class to prepare an initial version of a QMRF report.
  The XML output is in QMRF version 1.3 and can be finalized with the QMRF editor 2.0 (https://sourceforge.net/projects/qmrf/)
* QPRF:
  Provides a ruby OpenTox class to prepare an initial version of a QPRF (version 1.1) report in JSON or HTML.

## Usage
### QMRF
create a new QMRF report, add some content and show output:
```ruby
require "qsar-report"

# create a new report
report = OpenTox::QMRFReport.new

# add a title
report.value "QSAR_title", "My QSAR Title"

# change 6.2 'Available information for the training set' set inchi and smiles to Yes
report.change_attributes "training_set_data", {:inchi => "Yes", :smiles => "Yes"}

# add a publication to the publication catalog
report.change_catalog :publications_catalog, :publications_catalog_1, {:title => "MyName M (2016) My Publication Title, QSAR News, 10, 14-22", :url => "http://myqsarnewsmag.dom"}

# link/reference the publication to the report bibliography
report.ref_catalog :bibliography, :publications_catalog, :publications_catalog_1

# output
puts report.to_xml

# validate a report (as created above) against qmrf.xsd
report.validate
```
### QPRF
create a new QPRF report, add some content and show output:
```ruby
require "qsar-report"

# create a new QPRF report instance
report = OpenTox::QPRFReport.new

# Set Title of the report
report.Title = "My QPRF Report"

# Set Version
report.Version = "1"

# Set Date
report.Date = Time.now.strftime("%Y/%m/%d")

# Set the CAS number in chapter 1.1
report.value "1.1", "7732-18-5" # set CAS number for H²O

# print HTML version
puts report.to_html

# print formated JSON version
puts report.pretty_json

```

Installation
------------

  gem install qsar-report

Documentation
-------------
* [RubyDoc.info Code documentation](http://www.rubydoc.info/gems/qsar-report)
* For full information on QSAR reporting see: [JRC QSAR Model Database and QSAR Model Reporting Formats](https://eurl-ecvam.jrc.ec.europa.eu/databases/jrc-qsar-model-database-and-qsar-model-reporting-formats)

Copyright
---------
Copyright (c) 2016 Christoph Helma, Micha Rautenberg, Denis Gebele. See LICENSE for details.

Lazar QMRF Example
------------------
Here is an advanced example of the usage of the gem in lazar code:

{include:file:example/example.rb}



ChangeLog
* Add Tests.
* Change License text to GPL-3
v0.0.5
* add change_attributes method to edit QMRF 6.1-6.4 and 7.1-7.4
* fix catalog referencing code, adjust node search
v0.0.2
* add QPRF reporting
* rename gem qmrf-report to qsar-report
v0.0.1
* initial version

test/all.rb
["qmrf-report.rb", "qprf-report.rb"].each do |test|
  require_relative test
end


test/setup.rb
require 'minitest/autorun'
require_relative '../lib/qmrf-report.rb'
require_relative '../lib/qprf-report.rb'
include OpenTox
TEST_DIR ||= File.expand_path(File.dirname(__FILE__))
DATA_DIR ||= File.join(TEST_DIR,"data")

test/report.rb
require_relative "setup.rb"

class QMRFReportTest < MiniTest::Test

  def test_0_self
    puts "MiniTest start OK"
  end

  def test_1_base
    report = OpenTox::QMRFReport.new
    assert report
    assert_kind_of(OpenTox::QMRFReport, report)
  end

  def test_2_validate
    report = OpenTox::QMRFReport.new
    assert_empty report.validate
  end

  def test_3_write_some_values
    report = OpenTox::QMRFReport.new
    report.change_qmrf_tag "QSAR_title", "My QSAR Title"
    report.change_qmrf_tag "QSAR_models", "My QSAR Model"
    assert_equal report.get_qmrf_tag("QSAR_title"), "My QSAR Title"
    assert_equal report.get_qmrf_tag("QSAR_models"), "My QSAR Model"
    refute_equal report.get_qmrf_tag("QSAR_title"), "lazar"
  end

  def test_4_write_check
    report = OpenTox::QMRFReport.new
    report.change_qmrf_tag "QSAR_title", "My QSAR Title"
    report.change_qmrf_tag "QSAR_models", "My QSAR Model"
    assert_equal report.to_xml, File.read(File.join(File.join(DATA_DIR, "qmrf_t4.xml")))
  end

  def test_5_write_catalog
    report = OpenTox::QMRFReport.new
    report.change_catalog :software_catalog, :firstsoftware, {:name => "lazar", :contact => "in-silico toxicology gmbh", :url => "https://lazar.in-silico.ch", :description => "lazar toxicity predictions"}
    assert "in-silico toxicology gmbh", report.get_catalog_value(:software_catalog, :firstsoftware, :contact)
    assert "lazar", report.get_catalog_value(:software_catalog, :firstsoftware, :name)
    assert "https://lazar.in-silico.ch", report.get_catalog_value(:software_catalog, :firstsoftware, :url)
    assert "lazar toxicity predictions", report.get_catalog_value(:software_catalog, :firstsoftware, :description)
  end

  def test_6_check_catalog_exists
    report = OpenTox::QMRFReport.new
    assert_raises RuntimeError do
      report.catalog_exists? :noexist_catalog
    end
    assert report.catalog_exists? :software_catalog
  end
  
  def test_7_write_alot
    report = OpenTox::QMRFReport.new
    report.change_qmrf_tag "QSAR_title", "My QSAR Title"
    report.change_qmrf_tag "QSAR_models", "My QSAR Model"
    report.change_catalog :software_catalog, :firstsoftware, {:name => "lazar", :contact => "in-silico toxicology gmbh", :url => "https://lazar.in-silico.ch", :description => "lazar toxicity predictions"}
    report.change_catalog :publications_catalog, :publications_catalog_1, {:title => "MyName M (2016) My Publication Title, QSAR News, 10, 14-22", :url => "http://myqsarnewsmag.dom"}
    report.change_catalog :publications_catalog, :publications_catalog_2, {:title => "MyName M (2016) My Second Publication Title, Hornblower, 101ff.", :url => "http://hornblower.dom"}
    assert_equal report.to_xml, File.read(File.join(File.join(File.dirname(__FILE__),"data/qmrf_t7.xml")))
    report.change_catalog :publications_catalog, :publications_catalog_3, {:title => "MyName M (2016) My Third Publication Title, Somewhereelse, 43f.", :url => "http://somewhereelse.dom"}
    refute_equal report.to_xml, File.read(File.join(File.join(DATA_DIR, "qmrf_t7.xml")))
  end

  def test_8_ref_catalog_entry
    report = OpenTox::QMRFReport.new
    report.change_catalog :authors_catalog, :firstauthor, {:name => "Dr. My MyName", :url => "http://myauthor.dom", :email => "myauthor@myauthor.dom"}
    report.ref_catalog :qmrf_authors, :authors_catalog, :firstauthor
    report.change_catalog :authors_catalog, :secondauthor, {:name => "Dr. Mysec MysecName", :url => "http://myauthor.dom", :email => "myauthor@myauthor.dom"}
    report.ref_catalog :qmrf_authors, :authors_catalog, :secondauthor
    assert_equal report.to_xml, File.read(File.join(File.join(DATA_DIR, "qmrf_t8.xml")))
  end

end


example/example.rb
# require the qsar-report gem
require "qsar-report"

# generate a new report
report = OpenTox::QMRFReport.new

# lazar specific data resources for model information
model = Model::Lazar.find params[:id]
prediction_model = Model::Prediction.find_by :model_id => params[:id]
validation_template = File.join(File.dirname(__FILE__),"../views/model_details.haml") #detailed validation results from lazar-rest


# Start filling the QMRF report 

# QSAR identifier (title) 1.1
report.value "QSAR_title", "Model for #{prediction_model.species} #{prediction_model.endpoint}"

# Software coding the model 1.3
report.change_catalog :software_catalog, :firstsoftware, {:name => "lazar", :description => "lazar Lazy Structure- Activity Relationships", :number => "1", :url => "https://lazar.in-silico.ch", :contact => "Contact Email"}
report.ref_catalog :QSAR_software, :software_catalog, :firstsoftware

# Date of QMRF 2.1
report.value "qmrf_date", "#{Time.now.strftime('%d %B %Y')}"

# QMRF author(s) and contact details 2.1
report.change_catalog :authors_catalog, :firstauthor, {:name => "Christoph Helma", :affiliation => "in silico toxicology gmbh", :contact => "Contact Address", :email => "Contact Email", :number => "1", :url => "Web Page"}
report.ref_catalog :qmrf_authors, :authors_catalog, :firstauthor

# Model developer(s) and contact details 2.5
report.change_catalog :authors_catalog, :modelauthor, {:name => "Christoph Helma", :affiliation => "in silico toxicology gmbh", :contact => "Contact Address", :email => "Contact Email", :number => "1", :url => "Web Page"}
report.ref_catalog :model_authors, :authors_catalog, :modelauthor

# Date of model development and/or publication 2.6
report.value "model_date", "#{Time.parse(model.created_at.to_s).strftime('%Y')}"

# Reference(s) to main scientific papers and/or software package 2.7
report.change_catalog :publications_catalog, :publications_catalog_1, {:title => "Maunz, Guetlein, Rautenberg, Vorgrimmler, Gebele and Helma (2013), lazar: a modular predictive toxicology framework  ", :url => "http://dx.doi.org/10.3389/fphar.2013.00038"}
report.ref_catalog :references, :publications_catalog, :publications_catalog_1

# Reference(s) to main scientific papers and/or software package 2.7
report.change_catalog :publications_catalog, :publications_catalog_2, {:title => "Maunz A and Helma C (2008) Prediction of chemical toxicity with local support vector regression and activity-specific kernels. SAR & QSAR in Environmental Research 19 (5-6), 413-431", :url => "http://dx.doi.org/10.1080/10629360802358430"}
report.ref_catalog :references, :publications_catalog, :publications_catalog_2

# Species 3.1
report.value "model_species", prediction_model.species 

# Endpoint 3.2 
report.change_catalog :endpoints_catalog, :endpoints_catalog_1, {:name => prediction_model.endpoint, :group => ""}
report.ref_catalog :model_endpoint, :endpoints_catalog, :endpoints_catalog_1

# Endpoint Units 3.4
report.value "endpoint_units", "#{prediction_model.unit}"

# Type of model 4.1
report.value "algorithm_type", "#{model.class.to_s.gsub('OpenTox::Model::Lazar','')}"

# Explicit algorithm 4.2
report.change_catalog :algorithms_catalog, :algorithms_catalog_1, {:definition => "see Helma 2016 and lazar.in-silico.ch, submitted version: #{lazar_commit}", :description => "modified k-nearest neighbor classification with activity specific similarities, weighted voting and exhaustive enumeration of fragments and neighbors"}
report.ref_catalog :algorithm_explicit, :algorithms_catalog, :algorithms_catalog_1

# Descriptors in the model 4.3
report.change_catalog :descriptors_catalog, :descriptors_catalog_1, {:description => "all statistically relevant paths are used for similarity calculation", :name => "linear fragmens (paths)", :publication_ref => "", :units => "true/false (i.e. present/absent)"}
report.ref_catalog :algorithms_descriptors, :descriptors_catalog, :descriptors_catalog_1

# Descriptor selection 4.4
report.value "descriptors_selection", "statistical filter (chi-square with Yates correction)"

# Algorithm and descriptor generation 4.5
report.value "descriptors_generation", "exhaustive breadth first search for paths in chemical graphs (simplified MolFea algorithm)"

# Software name and version for descriptor generation 4.6
report.change_catalog :software_catalog, :software_catalog_2, {:name => "lazar, submitted version: #{lazar_commit}", :description => "simplified MolFea algorithm", :number => "2", :url => "https://lazar.in-silico.ch", :contact => "Contact Email"}
report.ref_catalog :descriptors_generation_software, :software_catalog, :software_catalog_2

# Chemicals/Descriptors ratio 4.7
report.value "descriptors_chemicals_ratio", "not applicable (classification based on activities of neighbors, descriptors are used for similarity calculation)"

# Some QMRF fields (tag contents) can have formatted HTML contents. 
# These contents have to have a complete HTML document structure with basic surrounding tags like <html><body>. 
# see Example below   

# Description of the applicability domain of the model 5.1
report.value "app_domain_description", "<html><head></head><body>
    <p>
      The applicability domain (AD) of the training set is characterized by 
      the confidence index of a prediction (high confidence index: close to 
      the applicability domain of the training set/reliable prediction, low 
      confidence: far from the applicability domain of the 
      trainingset/unreliable prediction). The confidence index considers (i) 
      the similarity and number of neighbors and (ii) contradictory examples 
      within the neighbors. A formal definition can be found in Helma 2006.
    </p>
    <p>
      The reliability of predictions decreases gradually with increasing 
      distance from the applicability domain (i.e. decreasing confidence index)
    </p>
  </body>
</html>"

# Method used to assess the applicability domain 5.2
report.value "app_domain_method", "see Helma 2006 and Maunz 2008"

# Software name and version for applicability domain assessment 5.3  
report.change_catalog :software_catalog, :software_catalog_3, {:name => "lazar, submitted version: #{lazar_commit}", :description => "integrated into main lazar algorithm", :number => "3", :url => "https://lazar.in-silico.ch", :contact => "Contact Email"}
report.ref_catalog :app_domain_software, :software_catalog, :software_catalog_3

# Limits of applicability 5.4
report.value "applicability_limits", "Predictions with low confidence index, unknown substructures and neighbors that might act by different mechanisms"

# Availability of the training set 6.1
report.change_attributes "training_set_availability", {:answer => "Yes"}

# Available information for the training set 6.2
report.change_attributes "training_set_data", {:cas => "Yes", :chemname => "Yes", :formula => "Yes", :inchi => "Yes", :mol => "Yes", :smiles => "Yes"}

# Data for each descriptor variable for the training set 6.3
report.change_attributes "training_set_descriptors", {:answer => "No"}

# Data for the dependent variable for the training set 6.4
report.change_attributes "dependent_var_availability", {:answer => "All"}

# Other information about the training set 6.5
report.value "other_info", "#{prediction_model.source}"

# Pre-processing of data before modelling 6.6
report.value "preprocessing", (model.class == OpenTox::Model::LazarRegression ? "-log10 transformation" : "none")

# Robustness - Statistics obtained by leave-many-out cross-validation 6.9
if prediction_model.crossvalidations
  crossvalidations = prediction_model.crossvalidations
  out = haml File.read(validation_template), :layout=> false, :locals => {:model => prediction_model}
  report.value "lmo",  out
end

# Mechanistic basis of the model 8.1
report.value "mechanistic_basis","<html><head></head><body>
  <p>
    Compounds with similar structures (neighbors) are assumed to have 
    similar activities as the query compound. For the determination of 
    activity specific similarities only statistically relevant subtructures 
    (paths) are used. For this reason there is a priori no bias towards 
    specific mechanistic hypothesis.
  </p>
</body>
</html>"

# A priori or a posteriori mechanistic interpretation 8.2
report.value "mechanistic_basis_comments","a posteriori for individual predictions"

# Other information about the mechanistic interpretation 8.3
report.value "mechanistic_basis_info","<html><head></head><body><p>Hypothesis about biochemical mechanisms can be derived from individual 
    predictions by inspecting neighbors and relevant fragments.</p>
    <p>Neighbors are compounds that are similar in respect to a certain 
    endpoint and it is likely that compounds with high similarity act by 
    similar mechanisms as the query compound. Links at the webinterface 
    prove an easy access to additional experimental data and literature 
    citations for the neighbors and the query structure.</p>
    <p>Activating and deactivating parts of the query compound are highlighted 
    in red and green on the webinterface. Fragments that are unknown (or too 
    infrequent for statistical evaluation are marked in yellow and 
    additional statistical information about the individual fragments can be 
    retrieved. Please note that lazar predictions are based on neighbors and 
    not on fragments. Fragments and their statistical significance are used 
    for the calculation of activity specific similarities.</p>"

# Reference all publications to the bibliography 
# Bibliography 9.2
report.ref_catalog :bibliography, :publications_catalog, :publications_catalog_1
report.ref_catalog :bibliography, :publications_catalog, :publications_catalog_2
report.change_catalog :publications_catalog, :publications_catalog_3, {:title => "Helma (2006), Lazy structure-activity relationships (lazar) for the prediction of rodent carcinogenicity and Salmonella mutagenicity.", :url => "http://dx.doi.org/10.1007/s11030-005-9001-5"}
report.ref_catalog :bibliography, :publications_catalog, :publications_catalog_3

# output
response['Content-Type'] = "application/xml"
return report.to_xml

lib/qprf-report.rb
require "json"
require "haml"

# OpenTox module
module OpenTox

  #Class for QPRF reporting.
  #
  #Provides a ruby OpenTox class to prepare an initial version of a QPRF report.
  #The QPRF output is in QPRF version 1.1  from May 2008
  #@example Report
  #  require "qsar-report"
  #  report = OpenTox::QPRFReport.new
  #  report.Title = "My QPRF Report"
  #  report.Version = "1"
  #  report.Date = Time.now.strftime("%Y/%m/%d")
  #  report.value "1.1", "7732-18-5" # set CAS number for H²O
  #  puts report.to_html

  class QPRFReport
    # QPRF JSON Template file
    TEMPLATE_FILE = File.join(File.dirname(__FILE__),"template/qprf.json")

    # QPRF MarkDown Template file
    MD_TEMPLATE_FILE = File.join(File.dirname(__FILE__),"template/qprf.haml")

    attr_accessor :json, :report

    # Open an existing QPRF json report
    # @param [String] file Name of the file
    def open file
      json = File.read("#{file}")
      @report = JSON.parse(json)
    end

    # Initialize a new report instance from QPRF template. With helper functions for Title, Version and Date
    def initialize
      json = File.read(TEMPLATE_FILE)
      @report = JSON.parse(json)

      attr_hash = {self.report['singleCalculations'] => ["Title", "Version", "Date"] }
      attr_hash.each_pair do |block, attributes|
        attributes.each do |attribute|
          define_singleton_method "#{attribute}" do
            return block[attribute]
          end
          define_singleton_method "#{attribute}=" do |val=nil|
            block[attribute] = val unless val.nil?
            return block[attribute]
          end
        end
      end

    end

    # Set or Get a value in the QPRF report
    #@example for CAS Number
    #  report = OpenTox::QPRFReport.new
    #  report.value "1.1", "7732-18-5"
    #
    # @param [String] chapter Name of the chapter - e.g.:  "1.1", "1.2", "1.3", "1.4", "1.5 General", "1.5 a.", "1.5 b.", "1.5 c.", "1.5 d.", "2.1" ...
    # @param [String] value Value to set. If not set the function returns the current value
    # @return [String]  returns Value
    def value chapter, value=nil
      case chapter
      when /^1\.\d*/
        block = "1. Substance"
      when /^2\.\d*/
        block = "2. General information"
      when /^3\.\d*/
        block = "3. Prediction"
      when /^4\.\d*/
        block = "4. Adequacy (Optional)"
      end
      @report["arrayCalculations"][block]['values'][chapter][1] = value unless value.nil?
      @report["arrayCalculations"][block]['values'][chapter][1]
    end

    # returns prettified JSON representation (QPRF JSON report) of report instance
    # @return [String] returns JSON
    def pretty_json
      JSON.pretty_generate(@report)
    end

    # Creates a HTML representation of the QPRF report
    # @return [String] returns HTML
    def to_html
      Haml::Engine.new(File.read(MD_TEMPLATE_FILE)).render @report
    end

  end
end

lib/qmrf-report.rb
require "nokogiri"

# OpenTox module
module OpenTox

  #Class for QMRF reporting. 
  #
  #Provides a ruby OpenTox class to prepare an initial version of a QMRF report. 
  #The XML output is in QMRF version 1.3 and can be finalized with the QMRF editor 2.0 (https://sourceforge.net/projects/qmrf/)  
  #@example Report
  #  require "qsar-report"
  #  report = OpenTox::QMRFReport.new
  #  report.value "QSAR_title", "My QSAR Title"
  #  report.change_attributes "training_set_data", {:inchi => "Yes", :smiles => "Yes"}
  #  report.change_catalog :publications_catalog, :publications_catalog_1, {:title => "MyName M (2016) My Publication Title, QSAR News, 10, 14-22", :url => "http://myqsarnewsmag.dom"}
  #  report.ref_catalog :bibliography, :publications_catalog, :publications_catalog_1
  #  puts report.to_xml

  class QMRFReport
    # QMRF XML Schema file 
    SCHEMA_FILE   = File.join(File.dirname(__FILE__),"template/qmrf.xsd")
    # QMRF XML Template file
    TEMPLATE_FILE = File.join(File.dirname(__FILE__),"template/qmrf.xml")
    # QMRF catalogs. Entries of a catalog can be referenced in certain tags.
    CATALOGS = ["software_catalog", "algorithms_catalog", "descriptors_catalog", "endpoints_catalog", "publications_catalog", "authors_catalog"]
    # QMRF XML tags with attributes to edit
    ATTRIBUTE_TAGS = ["training_set_availability", "training_set_data", "training_set_descriptors", "dependent_var_availability", "validation_set_availability", "validation_set_data", "validation_set_descriptors", "validation_dependent_var_availability"]
    
    attr_accessor :xml, :report

    # Open an existing QMRF xml report
    # @param [String] file Name of the file
    def open file
      xml = File.read("#{file}")
      @report = Nokogiri.XML(xml)
    end

    # Initialize a new report instance from qmrf template
    def initialize
      xml = File.read(TEMPLATE_FILE)
      @report = Nokogiri.XML(xml)
    end

    # returns XML representation (QMRF XML report) of report instance
    # @return [String] returns XML
    def to_xml
      @report.to_xml
    end

    # Get or Set a value
    # e.G.:
    #@example change the title
    #  report.value "QSAR_title", "Title of My QSAR"
    #  #changes the QSAR_title tag to:
    #  #<QSAR_title chapter="1.1" help="" name="QSAR identifier (title)">Title of My QSAR</QSAR_title>
    # @param [String] key Nodename e.g.: "QSAR_title"
    # @param [String] value Value to change. If not set the function returns the current value
    # @return [Error]  returns Error message if fails
    # @return [String] returns value    
    def value key, value=nil
      raise "Can not edit attribute #{key} directly. Edit the catalog with 'report.change_catalog(catalog, key, value)'." if ["QSAR_software","QSAR_Algorithm", ""].include? key
      t = @report.at_css key
      t.content = value unless value.nil?
      t.content
    end

    # Set attributes of an report XML tag.
    # Some of the QMRF XML tags have attributes to be edited. This applies to 6.1 to 6.4 and 7.1 to 7.4 see also: {OpenTox::QMRFReport::ATTRIBUTE_TAGS ATTRIBUTE_TAGS}.   
    # e.G. "Available information for the training set" at 6.2 of the report: 
    #  <training_set_data cas="Yes" chapter="6.2" chemname="Yes" formula="Yes" help="" inchi="Yes" mol="Yes" name="Available information for the training set" smiles="Yes"/>
    #@example change_attributes
    #  report.change_attributes "training_set_data", {:inchi => "Yes", :smiles => "Yes"}
    # @param [String] tagname Nodename e.g.: "training_set_data"
    # @param [Hash] valuehash Key-Value Hash of tag attributes to change.
    # @return [Error]  returns Error message if fails
    def change_attributes tagname, valuehash
      raise "Can not edit the attributes of tag: #{tagname}." unless ATTRIBUTE_TAGS.include? tagname
      tag = @report.at_css tagname
      valuehash.each do |key, value|
        tag.attributes["#{key}"].value = value
      end
    end

    # Change a catalog
    # @param [String] catalog Name of the catalog - One of {OpenTox::QMRFReport::CATALOGS CATALOGS}.
    # @param [String] id Single entry node in the catalog e.G.: "<software contact='mycontact@mydomain.dom' description="My QSAR Software " id="software_catalog_2" name="MySoftware" number="" url="https://mydomain.dom"/>
    # @param [Hash] valuehash Key-Value Hash with attributes for a single catalog node
    # @return [Error]  returns Error message if fails
    def change_catalog catalog, id, valuehash
      catalog_exists? catalog
      if @report.at_css("#{catalog}").at("[@id='#{id}']")
        valuehash.each do |key, value|
          @report.at_css("#{catalog}").at("[@id='#{id}']")["#{key}"]= value
        end
      else
        cat = @report.at_css("#{catalog}")
        newentry = Nokogiri::XML::Node.new("#{catalog.to_s.gsub(/s?_catalog/,'')}", self.report)
        newentry["id"] = id
        valuehash.each do |key, value|
          newentry["#{key}"] = value
        end
        cat << newentry
      end
    end

    # Set reference to a catalog entry.
    # e.g.: reference an author entry from authors_catalog to Chapter 2.2 QMRF authors
    #@example ref_catalog
    #  report.ref_catalog 'qmrf_authors', 'authors_catalog', 'firstauthor'
    # @param [String] chapter Name of the chapter to add the catalog reference. e.g.: qmrf_authors, model_authors, QSAR_software, ...
    # @param [String] catalog Name of the catalog. One of {OpenTox::QMRFReport::CATALOGS CATALOGS}.
    # @param [String] id entry node in the catalog
    def ref_catalog chapter, catalog, id
      catalog_exists? catalog
      if @report.at_css("#{catalog}").at("//*[@id='#{id}']")
        chap = @report.at_css("#{chapter}")
        if chap.at("[@idref='#{id}']").nil?
          newentry = Nokogiri::XML::Node.new("#{catalog.to_s.gsub(/s?_catalog/,'_ref')}", self.report)
          newentry["idref"] = id
          chap << newentry
        end
      else
        raise "catalog entry with id: #{id} do not exist."      
      end
    end

    # get an attribute from a catalog entry
    # @param [String] catalog Name of the catalog. One of {OpenTox::QMRFReport::CATALOGS CATALOGS}.
    # @param [String] id entry id in the catalog
    # @param [String] key returns value of a key in a catalog node
    # @return [String, false] returns value of a key in a catalog node or false if catalog entry do not exists.
    def get_catalog_value catalog, id, key
      catalog_exists? catalog
      if @report.at_css("#{catalog}").at("[@id='#{id}']")
        @report.at_css("#{catalog}").at("[@id='#{id}']")["#{key}"]
      else
        return false
      end
    end

    # Check if a catalog exists in this QMRF version
    # @param [String] catalog Catalog
    # @return [Error, true]  returns true or Error if a catalog do not exists. See also {OpenTox::QMRFReport::CATALOGS CATALOGS}.
    def catalog_exists? catalog
      raise "Unknown catalog: #{catalog}" unless CATALOGS.include? catalog.to_s
      true
    end

    # Validates a report instance against qmrf.xsd (XML Structure Definition)
    def validate
      xsd = Nokogiri::XML::Schema(File.read(SCHEMA_FILE))
      out = ""
      xsd.validate(@report).each do |error|
        out << error.message unless error.message == "Element 'algorithm', attribute 'publication_ref': '' is not a valid value of the atomic type 'xs:IDREF'." || error.message == "Element 'descriptor', attribute 'publication_ref': '' is not a valid value of the atomic type 'xs:IDREF'." 
        # @todo ignore case sensitivity error: error.message The value 'NO' is not an.Element of the set {'Yes', 'No'}.
      end
      return out
    end

  end
end